### ISSUE TITLE
Significant memory leak when TVP statements used with SQL Server

### ISSUE THREAD

**Issue by akinelden** [nanodbc/nanodbc#401]
## Environment

* nanodbc version: 2.14.0
* DBMS name/version: SQL Server 16.0.1110.1
* ODBC connection string: Driver={ODBC Driver 17 for SQL Server};Server=localhost;Database=TestDB;Encrypt=no;UID=sa;
* OS and Compiler: Windows Cmake VS2022
* CMake settings: default

## Actual behavior
When TVP is used with SQL Server, memory leak occurs because of circular reference between `statement_impl` and `table_valued_parameter_impl` classes. 

`table_valued_parameter_impl` holds a `statement` (which means it holds a `std::shared_ptr<statement_impl>`) in nanodbc.cpp line 3025:
```
class table_valued_parameter::table_valued_parameter_impl
{
...
private:
    class statement stmt_;
...
};
```
`statement_impl` holds a map of `table_valued_parameter` in nanodbc.cpp line 2400 and it's never cleared explicitly:
```
class statement::statement_impl
{
...
#ifndef NANODBC_DISABLE_MSSQL_TVP
    std::map<short, table_valued_parameter> tvp_data_;
    bool open_tvp_;
#endif
};
```

## Expected  behavior
`statement_impl::tvp_data_` should be cleaned inside `statement_impl::close()` function. However, this doesn't prevent the memory leak if `statement::close()` is never called (e.g. directly destructed when goes out of scope). Therefore, `table_valued_parameter_impl` should hold `std::weak_ptr<statement_impl>` instead of `std::shared_ptr`.

## Minimal Working Example
Include `<iostream>` and add `std::cout` to constructor and destructors of `statement_impl` and `table_valued_parameter_impl` classes. Run the table_valued_parameter.cpp example in the repository. When the `statement` and `table_valued_parameter` goes out of scope, you will not see the destructor messages.

Add a message to nanodbc.cpp line 1615
```
    ~statement_impl() noexcept
    {
        std::cout << "statement_impl is destructed" << std::endl;
        ...
    }
```

Add a message to nanodbc.cpp line 2584
```
    ~table_valued_parameter_impl() noexcept
    {
        std::cout << "table_valued_parameter_impl is destructed" << std::endl;
        ...
    }
```

Add constructor messages similarly. Below are the screenshots from console output. You can see that one `statement_impl` and the `table_valued_parameter_impl` are not destructed.

Console output before starting TVP statement:
![nanodbc-example-start](https://github.com/nanodbc/nanodbc/assets/43653687/5cbb5a02-07fe-4321-b792-d478bd793cab)

Console output after TVP statement is completed and goes out of scope:
![nanodbc-example-end](https://github.com/nanodbc/nanodbc/assets/43653687/91fba5a2-4571-49ab-9cd7-cb522b478084)

**Comment by akinelden** [comment]
I solved the problem in our project by doing aforementioned changes and created a PR for the issue.

**Comment by mloskot** [comment]
Thanks for the report. 
/cc @sphawk as the author of the feature may be interested too
